File: ios/Packages/GemstoneServices/Sources/Stores/AddressStore.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import struct Gemstone.AddressName
import typealias Gemstone.Chain
import struct Gemstone.GemAddressNameUpdate
import protocol Gemstone.GemAddressStore
import GemstonePrimitives
import Primitives
import Store

public final class GemstoneAddressStore: GemAddressStore, @unchecked Sendable {
    private let store: AddressStore

    public init(store: AddressStore) {
        self.store = store
    }

    public func getAddressName(chain: Gemstone.Chain, address: String) throws -> Gemstone.AddressName? {
        try store.getAddressName(chain: Primitives.Chain(id: chain), address: address).map { $0.toGem() }
    }

    public func saveAddressNames(updates: [GemAddressNameUpdate]) async throws {
        try store.updateAddressNames(
            updates.map { AddressNameUpdate(name: $0.name.toPrimitives(), replacesTypes: $0.replacesTypes.map { $0.toPrimitives() }) },
        )
    }

    public func deleteAddressNames(names: [Gemstone.AddressName]) async throws {
        try store.deleteAddressNames(names.map { $0.toPrimitives() })
    }
}

File: android/data/services/gemstone/src/main/kotlin/com/gemwallet/android/data/services/gemstone/stores/AddressStore.kt
package com.gemwallet.android.data.services.gemstone.stores

import com.gemwallet.android.data.services.store.database.AccountsDao
import com.gemwallet.android.data.services.store.database.AddressesDao
import com.gemwallet.android.data.services.store.database.WalletsDao
import com.gemwallet.android.data.services.store.database.entities.AddressNameUpdate
import com.gemwallet.android.data.services.store.database.entities.toDTO
import com.gemwallet.android.data.services.store.database.entities.toRecord
import com.gemwallet.android.ext.requireChain
import com.gemwallet.android.ext.toGem
import com.gemwallet.android.ext.toPrimitives
import kotlinx.coroutines.flow.first
import uniffi.gemstone.GemAddressNameUpdate
import uniffi.gemstone.GemAddressStore

class GemstoneAddressStore(private val addressesDao: AddressesDao, private val accountsDao: AccountsDao, private val walletsDao: WalletsDao) : GemAddressStore {

    override suspend fun getAddressName(chain: String, address: String): uniffi.gemstone.AddressName? {
        val record = addressesDao.get(chain.requireChain(), address) ?: return null
        val walletImageUrl = accountsDao.getByAddress(record.chain, address).firstNotNullOfOrNull { account -> walletsDao.getById(account.walletId).first()?.imageUrl }
        return record.copy(imageUrl = record.imageUrl ?: walletImageUrl).toDTO().toGem()
    }

    override suspend fun saveAddressNames(updates: List<GemAddressNameUpdate>) {
        if (updates.isEmpty()) return
        addressesDao.updateNames(
            updates.map { update ->
                AddressNameUpdate(update.name.toPrimitives().toRecord(), update.replacesTypes.map { it.toPrimitives() })
            },
        )
    }

    override suspend fun deleteAddressNames(names: List<uniffi.gemstone.AddressName>) {
        if (names.isEmpty()) return
        addressesDao.deleteNames(names.map { it.toPrimitives() }.toRecord())
    }
}
